### ISSUE TITLE
minimum_cut returns correct value, valid partition, but different from what it claims to return

### ISSUE THREAD

**Issue by alabarre** [networkx/networkx#5239]
I see two open issues related to `minimum_cut` but I have just encountered yet another problem. The function claims to return a partition `(left, right)` where `left` contains all nodes reachable from the source in the residual network and `right` contains the rest, but the following instance shows that the code behaves differently:

    >>> import networkx as nx
    >>> G = nx.DiGraph()
    >>> for u, v, c in [(35, 36, 2), (36, 37, 2), (37, 39, 2), (39, 40, 2), (40, 41, 1), (41, 46, 1), (46, 47, 1)]:
    ...     G.add_edge(u, v, capacity=c)
    >>> nx.minimum_cut(G, 35, 47)
    (1, ({35, 36, 37, 39, 40, 41, 46}, {47}))

Although the cut is minimum, the flow on `(40, 41)` must have value 1, and therefore node `41` is not reachable from 35 in the residual network. So `minimum_cut` should return `({35, 36, 37, 39, 40}, {41, 46, 47})`.

(using networkx version 2.4, Python 3.8.10 on Ubuntu 20.04.3)

**Comment by dschult** [comment]
Why do you say the node `41` is unreachable from node `35` in the residual network? If the cut is made on edge `(46, 47)`, the cut has value `1` and reachable nodes are as the coe returns. Perhaps in this case there are more than one minimum cut?  But maybe I'm missing something...

**Comment by alabarre** [comment]
The maximum flow has value 1 and assigns a flow of 1 on every arc. Therefore, the corresponding residual network has the following arcs:

-  (35, 36, 1), (36, 37, 1), (37, 39, 1), (39, 40, 1) (we've used 1/2 units, so those "forward" residual capacities have a value of 1);
-  (36, 35, 1), (37, 36, 1), (39, 37, 1), (40, 39, 1) (same reason: we've pushed a flow of 1, so those "backward" capacities have a value of 1);
-  (41, 40, 1), (46, 41, 1), (47, 46, 1) (same weight assignment as above)
-  and no other arc since arcs with unit capacity are saturated.

So there is no directed path from 35 to 41 in this residual network. Or did I misinterpret anything?

(yes, there are other minimum cuts --- see the one I gave above)

**Comment by dschult** [comment]
A minimum cut is a cut with the smallest total weight of the edges which if removed would disconnect the source from the sink.  So if we cut (46, 47, 1) the source `35` and target `47` are disconnected and the total edge weights cut is 1.

I think your question can be boiled down to:  Should the unreachable node set be the set of nodes that *can* reach the target after the cut? Or should it be the set of nodes that can't be reached from the source node?  

The code clearly computes the nodes that *can* reach the target, while the comment just before the code says to compute the nodes that can't be reached from the source. Looking at the "blame" history, this can be traced to #1102 where there is[ a commit](https://github.com/networkx/networkx/commit/644a7b6a378430310cb323f17648943ae70b7a31) that has in it the commit message:  `Node partition should be determined by reachability to the sink.`

From my perspective, the cut can be `(46,47)` which gives the existing return value. But it can also be `(40, 41)` which gives the proposed return value. But either way, the comment within the code is old and should be removed or changed. And maybe the code should be changed. And maybe the doc_string should be changed.

Again, I believe the current code is correct -- but perhaps not matching the documentation and perhaps not producing what is most natural.

**Comment by alabarre** [comment]
We agree on the definitions and on the fact that both cuts are minimum.

My issue is with the behaviour of the code with respect to the comment (again, in version 2.4, which seems to be the same as version 2.6): "(...) **reachable** and non reachable nodes **from source** in the residual network form the node partition that defines the minimum cut.". 

Maybe I'm parsing this the wrong way, but reading this leads me to expect the cut `(S, T)` to be defined by "`S` = source + all nodes that can be reached from the source in the residual network and `T` = V \ S" (as is standard); whereas the code seems to compute "`T` = all nodes that allow us to reach the target and `S` = V \ T". 

Thanks for linking to the commit, which allows me to further clarify my point of view: the version previous to that commit was correct with respect to the comment I quoted above.

**Comment by dschult** [comment]
Yes -- I agree.  The comment described the code prior to that commit.  It should have been changed when the code was changed -- OR -- the code shouldn't have been changed.  

@ysitu do you recall why it is better to partition the nodes based on those that cannot reach the sink?
(this can be traced to #1102 where there is a commit that has in it [the commit message](https://github.com/networkx/networkx/commit/644a7b6a378430310cb323f17648943ae70b7a31): 
Node partition should be determined by reachability to the sink.)
